datasets/ml_spoken_words/ml_spoken_words.py

```diff
@@ -0,0 +1,245 @@
+    "ky",
+    "lt",
+    "lv",
+    "mn",
+    "mt",
+    "nl",
+    "or",
+    "pl",
+    "pt",
+    "rm-sursilv",
+    "rm-vallader",
+    "ro",
+    "ru",
+    "rw",
+    "sah",
+    "sk",
+    "sl",
+    "sv-SE",
+    "ta",
+    "tr",
+    "tt",
+    "uk",
+    "vi",
+    "zh-CN",
+]
+
+
+class MlSpokenWordsConfig(datasets.BuilderConfig):
+    """BuilderConfig for MlSpokenWords."""
+
+    def __init__(self, *args, languages, **kwargs):
+        """BuilderConfig for MlSpokenWords.
+        Args:
+            languages (:obj:`Union[List[str], str]`): language or list of languages to load
+            **kwargs: keyword arguments forwarded to super.
+        """
+        super().__init__(
+            *args,
+            name="+".join(languages) if isinstance(languages, list) else languages,
+            **kwargs,
+        )
+        self.languages = languages if isinstance(languages, list) else [languages]
+
+
+class MlSpokenWords(datasets.GeneratorBasedBuilder):
+    """
+    Multilingual Spoken Words Corpus is a large and growing audio dataset of spoken
+    words in 50 languages collectively spoken by over 5 billion people, for academic
+    research and commercial applications in keyword spotting and spoken term search,
+    licensed under CC-BY 4.0. The dataset contains more than 340,000 keywords,
+    totaling 23.4 million 1-second spoken examples (over 6,000 hours).
+    """
+
+    VERSION = _VERSION
+    BUILDER_CONFIGS = [MlSpokenWordsConfig(languages=[lang], version=_VERSION) for lang in _LANGUAGES]
+    BUILDER_CONFIG_CLASS = MlSpokenWordsConfig
+
+    def _info(self):
+        features = datasets.Features(
+            {
+                "file": datasets.Value("string"),
+                "is_valid": datasets.Value("bool"),
+                "language": datasets.ClassLabel(names=self.config.languages),
+                "speaker_id": datasets.Value("string"),
+                "gender": datasets.ClassLabel(names=_GENDERS),
+                "keyword": datasets.Value("string"),  # seems that there are too many of them (340k unique keywords)
+                "audio": datasets.Audio(sampling_rate=48_000),
+            }
+        )
+        return datasets.DatasetInfo(
+            description=_DESCRIPTION,
+            features=features,
+            homepage=_HOMEPAGE,
+            license=_LICENSE,
+            citation=_CITATION,
+        )
+
+    def _split_generators(self, dl_manager):
+        splits_archive_path = [dl_manager.download(_SPLITS_URL.format(lang=lang)) for lang in self.config.languages]
+        download_audio = partial(_download_audio_archives, dl_manager=dl_manager)
+
+        return [
+            datasets.SplitGenerator(
+                name=datasets.Split.TRAIN,
+                gen_kwargs={
+                    "audio_archives": [download_audio(split="train", lang=lang) for lang in self.config.languages],
+                    "splits_archive": [dl_manager.iter_archive(path) for path in splits_archive_path],
```

Comment:
I think currently this would cause an issue when shuffling in streaming mode. Indeed shuffling does shuffle the order of the lists that are passed to _split_generators because each list would be shuffled differently. I can fix this by making them be shuffled the exact same way.

The alternative would be to use a single list of `zip(audio_archives, splits_archive)` but I feel like it would make more sense if `datasets` takes care of that naturally

Reply:
but is it a problem in this specific case? here a single splits_archive contains three csv files with metadata for train / val /test splits, so I iterate over them to take only one csv that corresponds to the current split. (I made them tar.gz archives because some csv files for large languages like English can be up to 1Gb in size)

Reply:
> you mean zip(audio_archives, splits_archive) should be passed directly to the _generate_examples, as a single parameter, right?

Yup that would work. Anyway I opened this PR to fix that for every dataset, because I don't like that it could cause silent issues:

https://github.com/huggingface/datasets/issues/3744

Reply:
Feel free to zip them and pass them as a single parameter for now, so that users that have an older version of the lib don't have issues. But in the future we should be able to have them in two separate lists :)

Reply:
heh it looks much uglier that way :( note that I can't refactor `[dl_manager.iter_archive(path) for path in splits_archive_path]` to a single variable as it would lead to stop iteration and the next splits could not be prepared

Reply:
maybe it wasn't a good idea to put all the splits csv metadata files into a single tar.gz. archive... my motivation was that since the dataset is already very large I should compress all that I can compress 😈

Reply:
@lhoestq I checked the shuffling PR https://github.com/huggingface/datasets/pull/3746  and it seems that we misunderstood each other. I guess in my case here I actually don't need zipping archives since they are not aligned (each `splits_archive` contains only three files: `train.csv`, `val.csv`, `test.csv`, each audio_archive contains all the audio files) and their order actually doesn't matter. agree? can I roll back?

Reply:
Oh ok my bad ! Sure feel free to roll back